### PR TITLE
JSON-formatted output

### PR DESCRIPTION
**Pull request by mleinhos**
This PR features JSON output for DRAKVUF. It relies on the existing dependency json-c and requires no configuration/build alterations. It has been tested against several tens of millions of events.

### CODE REVIEW

File: src/libdrakvuf/libdrakvuf.h [hunk 1]
```diff
@@ -115,6 +115,7 @@ extern "C" {
 #include <libvmi/libvmi.h>
 #include <libvmi/events.h>
 #include <json-c/json.h>
+#include <string.h>
```
Comment (tklengyel) [inline src/libdrakvuf/libdrakvuf.h]: Do we really need this new include?

File: src/main.cpp [hunk 2]
```diff
@@ -182,7 +182,7 @@ int main(int argc, char** argv)
                 "\t -j, --injection-timeout <seconds>\n"
                 "\t                           Injection timeout (in seconds, 0 == no timeout)\n"
                 "\t -t <timeout>              Timeout (in seconds)\n"
-                "\t -o <format>               Output format (default or csv)\n"
+                "\t -o <format>               Output format (default, csv, or json)\n"
```
Comment (tklengyel) [inline src/main.cpp]: We seem to have missed adding `kv` previously to the list, please do add that too.

File: src/run_drakvuf-async.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
+
```
Comment (tklengyel) [inline src/run_drakvuf-async.py]: Please add license blurb

File: src/run_drakvuf.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/env python3
+
```
Comment (tklengyel) [inline src/run_drakvuf.py]: Please add license blurb

File: src/libdrakvuf/libdrakvuf.h (hunk 1 above)
Comment (mleinhos) [reply src/libdrakvuf/libdrakvuf.h]: No - I'll move it to `drakvuf.c`

**Comment by tklengyel** [comment]
You also seem to be updating the Xen submodule, I don't think that belongs in this PR.

File: src/plugins/exmon/exmon.cpp [hunk 5]
```diff
@@ -163,16 +163,27 @@ static const char* offset_names[__OFFSET_MAX][2] =
 
 static void print_program_info(uint8_t previous_mode, char const* user_format, drakvuf_trap_info_t* info)
 {
+    exmon* e = (exmon*)info->trap->data;
     if (previous_mode == 1)
     {
         if (info->proc_data.base_addr)
         {
-            printf(user_format, info->proc_data.pid, info->proc_data.ppid, info->proc_data.name);
+            char * escaped_pname = NULL;
```
Comment (mleinhos) [inline src/plugins/exmon/exmon.cpp]: This should be initialized to `info->proc_data.name`

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

**Comment by tklengyel** [comment]
The Xen submodule is still being updated in this PR

**Comment by mleinhos** [comment]
Right. I haven't completed my corrections yet.

**Comment by mleinhos** [comment]
You're welcome to pull this in, but I haven't tested it to my own satisfaction yet. Windows changed the kernel of my test VM and I'm having trouble getting the updated PDB. I'll do further testing once I've gotten this figured out, but it could be a couple days with my other obligations.

**Comment by tklengyel** [comment]
Travis is still complaining about style so that would also have to be fixed. For test VMs its probably best to pull the network interface altogether to avoid it being updated :) Microsoft can take a couple days to push their pdbs for new kernels.

**Comment by mleinhos** [comment]
Can you point me to the style docs so I can fix that? Good idea on the
networking.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by mleinhos** [comment]
I'm working through the formatting now. It's a pain. A tool in the repo
that formats the code correctly would be helpful in this case....

**Comment by tklengyel** [comment]
You just have to run
 ```
export ARTISTIC_STYLE_OPTIONS=".astylerc" && astyle `find . | egrep "\.c|\.h"`

File: src/libdrakvuf/drakvuf.c [hunk 6]
```diff
@@ -672,6 +674,112 @@ unicode_string_t* drakvuf_read_wchar_string(vmi_instance_t vmi, const access_con
...
+ * \A\B----c (c shifted right 3)
+ * \A---B--c (B shifted right 2)
+ * --A--B--c (A shifted right 1)
+ * \\A\\B\\c (output string: NULLs filled in with \ chars)
+ *
+ * TODO: Handle non-ASCII character sets
+ *
+ * See https://www.freeformatter.com/json-escape.html
+ */
+char* drakvuf_escape_str(const char* input)
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c]: Instead of cooking up our own thing here, why not use `g_strescape`?

**Comment by mleinhos** [comment]
That would help us only partially. We're going to need our own solution
because g_strescape() and JSON handle non-ASCII values differently. See
https://json.org/ for details. Our custom solution here still needs work to
make it fully JSON compliant, but it covers the bulk of the cases I've seen
on my test VM.

**Comment by tklengyel** [comment]
Have you considered using something like https://github.com/nlohmann/json?

**Comment by mleinhos** [comment]
I have not. Are you willing to assume another external dependency for
DRAKVUF?  Another option might be https://github.com/zserge/jsmn, I need to
read up on it.

As things are now, I've already updated the plugins to support JSON and
would prefer not to rewrite all that to another API. However, it seems
reasonable, at a minimum, to rewrite drakvuf_escape_str() to use a real
JSON library.

How would you like to move forward?

Kind regards,
--Matt

**Comment by tklengyel** [comment]
Adding an another (optional) dependency would be completely fine as long as the license is compatible (MIT is fine in this case). I would much rather have a standard and separately maintained dependency then having custom stuff internal that we have to maintain going forward.

**Comment by tklengyel** [comment]
Check out https://github.com/nlohmann/json#serialization--deserialization, seems like would be a pretty simple change to what you have right now.

**Comment by tklengyel** [comment]
This also looks like a good candidate: https://github.com/Tencent/rapidjson, also supports UTF-16 which might be handy.

**Comment by mleinhos** [comment]
Your suggestions are in C++. I'm assuming you don't foresee the need to
emit a JSON-formatted string from C?

**Comment by tklengyel** [comment]
Yes, all plugins are in C++ so that would be fine.

**Comment by mleinhos** [comment]
Do you have in mind a specific file where the new C++ function would reside?

**Comment by tklengyel** [comment]
Not sure I follow, what new C++ function? I figure you can include the json header (when enabled via  configure) in `plugins/private.h` and then each plugin can use it as it sees fit (inside an `#ifdef ENABLE_JSON_OUTPUT` block). The headers wouldn't have to be included in drakvuf itself, both rapidjson and nlohmann json are available on debian/ubuntu as packages: `nlohmann-json3-dev` `rapidjson-dev`

**Comment by mleinhos** [comment]
Ah, ok. I had a different solution in mind that would minimize changing the plugins further. I'm have that solution completed soon for your review.

**Comment by mleinhos** [comment]
You've got the essence of my proposed change with the external JSON repo - I'll have another adjustment or two forthcoming. The build is failing because the new submodule isn't being fetched by the testing script.

**Comment by tklengyel** [comment]
I don't think we need a submodule for these headers, I much prefer if the are installed from the distributions package manager.

**Comment by tklengyel** [comment]
No, this PR as-is is not mergable - all CI gates fail. You are mixing C++ code into `libdrakvuf` which is exclusively C. The json output is only used plugins so it needs to be properly put in there. Since the headers required to build with nlohmann json are available as pacakges, there is no need to pull it in as a submodule. The package has to be installed for the Travis script as well to pass CI. Furthermore, I would like to see this as an optional feature, so you have to check if the json header is available in `configure.ac` and only enable it if it is.

File: src/libdrakvuf/Makefile.am [hunk 7]
```diff
@@ -128,5 +128,5 @@ endif
 endif
 
 noinst_LTLIBRARIES= libdrakvuf.la
-libdrakvuf_la_SOURCES= $(h_sources) $(c_sources)
+libdrakvuf_la_SOURCES= $(h_sources) $(c_sources) $(cpp_sources)
```
Comment (mleinhos) [inline src/libdrakvuf/Makefile.am]: revert this line back to master

File: src/plugins/bsodmon/bsodmon.cpp [hunk 8]
```diff
@@ -121,6 +121,7 @@ static event_response_t hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
     uint64_t code = 0;
     uint64_t params[4] = { 0 };
     const char* bugcheck_name = "UNKNOWN_CODE" ;
+    char* escaped_pname = NULL;
```
Comment (mleinhos) [inline src/plugins/bsodmon/bsodmon.cpp]: all results of drakvuf_ecape_str() should be gchar* for correctness

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@mleinhos please rebase this PR on the latest master. Afterwards I'm OK with merging

File: src/libinjector/injector.c [hunk 9]
```diff
@@ -1521,20 +1534,14 @@ int injector_start_app(
...
-            injector->error_code = err;
-            injector->error_string = err_str;
             print_injection_info(format, file, injector);
         }
     }
 
     rc = injector->rc;
-    PRINT_DEBUG("Finished with injection. Ret: %i\n", rc);
+    PRINT_DEBUG("Finished with injection. Ret: %i. Error: %s(%d)\n", rc,
+                injector->error_string ?: "OK", injector->error_code);
```
Comment (tklengyel) [inline src/libinjector/injector.c]: Are these changes intentional in `injector.c`? I just noticed them and they don't like like they belong in this PR

Comment (mleinhos) [reply src/libinjector/injector.c]: They are not -- I didn't add that change myself. I'll see if I can figure out what happened there.

Comment (mleinhos) [reply src/libinjector/injector.c]: I see that change in both our branches (your master and my json). It was introduced by zodeak on 5 Feb, so it must have been pulled in from the rebase. That would imply the left side of this diff shows the repo state prior to my branch, rather than the current state of master.

Comment (tklengyel) [reply src/libinjector/injector.c]: OK, yes, it matches what's in the latest master. I recommend instead of merging master into your branch using `git rebase`, it usually results in cleaner results.

**Comment by tklengyel** [comment]
Thanks again for the contribution!

**Comment by mleinhos** [comment]
You're welcome! I'm hopeful it will make DRAKVUF more usable within big
data applications, e.g. kafka.
